mor/mor/feel/feelmor/pbdw.hpp

```diff
@@ -38,15 +38,326 @@
+    if( toComplete )
+        ids = idsWith(sensors);
+    else
+        ids = sensors;
+    vectorN_type yobs2 = vectorN_type::Zero(ids.size());
+    yobs2.head(ids.size() - this->M_N) = yobs;
+    vectorN_type vn = M_matrix(ids,ids).colPivHouseholderQr().solve(yobs2);
+    return vn;
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::outputs(vectorN_type const& yobs, std::vector<int> const& sensors, bool toComplete) const
+{
+   std::vector<int> ids;
+    if( toComplete )
+        ids = idsWith(sensors);
+    else
+        ids = sensors;
+    vectorN_type coeffs = this->online(yobs, ids, false);
+    vectorN_type vn = M_F(Eigen::all, ids)*coeffs;
+    return vn;
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::online(vectorN_type const& yobs, std::vector<std::string> const& sensors) const
+{
+    std::vector<int> ids = idsWith(sensors);
+    return this->online(yobs, ids, false);
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::outputs(vectorN_type const& yobs, std::vector<std::string> const& sensors) const
+{
+    std::vector<int> ids = idsWith(sensors);
+    return this->outputs(yobs, ids, false);
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::onlineWithout(vectorN_type const& yobs, std::vector<int> const& sensors) const
+{
+    auto ids = idsWithout(sensors);
+    return online(yobs, ids, false);
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::outputsWithout(vectorN_type const& yobs, std::vector<int> const& sensors) const
+{
+    auto ids = idsWithout(sensors);
+    return outputs(yobs, ids, false);
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::onlineWithout(vectorN_type const& yobs, std::vector<std::string> const& sensors) const
+{
+    auto ids = idsWithout(sensors);
+    return online(yobs, ids, false);
+}
+
+typename PBDWOnline::vectorN_type
+PBDWOnline::outputsWithout(vectorN_type const& yobs, std::vector<std::string> const& sensors) const
+{
+    auto ids = idsWithout(sensors);
+    return outputs(yobs, ids, false);
+}
+
+void
+PBDWOnline::loadDB( std::string const& filename, crb::load l )
+{
+    if( !fs::exists(filename) )
+        return;
+
+    fs::ifstream ifs( filename );
+    if ( ifs )
+    {
+        Feel::cout << "loading DB at " << filename << std::endl;
+        boost::archive::binary_iarchive ia( ifs );
+        if( l > crb::load::none )
+        {
+            ia >> this->M_N;
+            ia >> this->M_M;
+            this->M_matrix.resize(this->M_M+this->M_N, this->M_M+this->M_N);
+            ia >> this->M_matrix;
+            ia >> this->M_Nl;
+            this->M_F.resize(this->M_Nl, this->M_M+this->M_N);
+            ia >> this->M_F;
+            this->M_sensorNames.reserve(M_M);
+            ia >> this->M_sensorNames;
+            ia >> this->M_outputNames;
+        }
+    }
+}
+
 /**
  * Class that implements Parameterized Background Data-Weak method.
  * @tparam Reduced basis space
  */
 template<typename RBSpace>
-class PBDW : CRBDB
+class PBDW : public PBDWOnline
```

Comment:
why depend on PBDWOnline ? why not a member instead ?

Reply:
I don't see why PBDW should own PBDWOnline. It is just a layer that add capability to do the offline phase.